spec/ddtrace/contrib/faraday/middleware_spec.rb

```diff
@@ -0,0 +1,165 @@
+      builder.use(:ddtrace, middleware_options)
+      builder.adapter(:test) do |stub|
+        stub.get('/success') { |_| [200, {}, 'OK'] }
+        stub.post('/failure') { |_| [500, {}, 'Boom!'] }
+        stub.get('/not_found') { |_| [404, {}, 'Not Found.'] }
+      end
+    end
+  end
+
+  let(:middleware_options) { {} }
+  let(:configuration_options) { { tracer: tracer } }
+
+  let(:request_span) do
+    tracer.writer.spans(:keep).find { |span| span.name == Datadog::Contrib::Faraday::NAME }
+  end
+
+  before(:each) do
+    Datadog.configure do |c|
+      c.use :faraday, configuration_options
+    end
+  end
+
+  context 'when there is no interference' do
+    subject!(:response) { client.get('/success') }
+
+    it do
+      expect(response).to be_a_kind_of(::Faraday::Response)
+      expect(response.body).to eq('OK')
+      expect(response.status).to eq(200)
+    end
+  end
+
+  context 'when there is successful request' do
+    subject!(:response) { client.get('/success') }
+
+    it do
+      expect(request_span).to_not be nil
+      expect(request_span.service).to eq(Datadog::Contrib::Faraday::SERVICE)
+      expect(request_span.name).to eq(Datadog::Contrib::Faraday::NAME)
+      expect(request_span.resource).to eq('GET') 
+      expect(request_span.get_tag(Datadog::Ext::HTTP::METHOD)).to eq('GET') 
+      expect(request_span.get_tag(Datadog::Ext::HTTP::STATUS_CODE)).to eq('200') 
+      expect(request_span.get_tag(Datadog::Ext::HTTP::URL)).to eq('/success') 
+      expect(request_span.get_tag(Datadog::Ext::NET::TARGET_HOST)).to eq('example.com') 
+      expect(request_span.get_tag(Datadog::Ext::NET::TARGET_PORT)).to eq('80') 
+      expect(request_span.span_type).to eq(Datadog::Ext::HTTP::TYPE) 
+      expect(request_span.status).to_not eq(Datadog::Ext::Errors::STATUS)
+    end
+  end
+
+  context 'when there is a failing request' do
+    subject!(:response) { client.post('/failure') }
+
+    it do
+      expect(request_span.service).to eq(Datadog::Contrib::Faraday::SERVICE)
+      expect(request_span.name).to eq(Datadog::Contrib::Faraday::NAME)
+      expect(request_span.resource).to eq('POST')
+      expect(request_span.get_tag(Datadog::Ext::HTTP::METHOD)).to eq('POST')
+      expect(request_span.get_tag(Datadog::Ext::HTTP::URL)).to eq('/failure')
+      expect(request_span.get_tag(Datadog::Ext::HTTP::STATUS_CODE)).to eq('500')
+      expect(request_span.get_tag(Datadog::Ext::NET::TARGET_HOST)).to eq('example.com')
+      expect(request_span.get_tag(Datadog::Ext::NET::TARGET_PORT)).to eq('80')
+      expect(request_span.span_type).to eq(Datadog::Ext::HTTP::TYPE)
+      expect(request_span.status).to eq(Datadog::Ext::Errors::STATUS)
+      expect(request_span.get_tag(Datadog::Ext::Errors::TYPE)).to eq('Error 500')
+      expect(request_span.get_tag(Datadog::Ext::Errors::MSG)).to eq('Boom!')
+    end
+  end
+
+  context 'when there is a client error' do
+    subject!(:response) { client.get('/not_found') }
```

Comment:
Nit: I feel a `before` hook would be more explicit. I also think that it's weird to call `client.get` your test subject if you're actually testing the side effect of it, rather than the return value itself. Feel free to ignore it though 😛

Reply:
True that we're asserting a side effect, but I might argue what we're testing is the operation that's supposed to produce the trace under test, which you could reasonably say is the "subject" of the test.